### ISSUE TITLE
Bug in initialisation of options grouped in classes (ArgGroup)

### ISSUE THREAD

**Issue by eqvis** [remkop/picocli#1685]
Hi Remko,

I'm using:
[picocli INFO] Picocli version: 4.6.3, JVM: 17.0.1 (Eclipse Adoptium OpenJDK 64-Bit Server VM 17.0.1+12), OS: Windows 10 10.0 amd64
With JLine: 3.21.0.

We have some options grouped in reusable classes. This seems to work sometimes, but then, in combination with arguments it does not work as expected. As an example I have two exclusive options `-a` and `-b` and I wrote a test that calls the command 4 times and prints the following:

Executing line: 'reuse testAddDomain'
a: 'null' | b: 'null'
Executing line: 'reuse testAddDomain -a A'
a: 'A' | b: 'null'
Executing line: 'reuse testAddDomain -b B'
a: 'null' | b: 'B'
Executing line: 'reuse testAddDomain'
a: 'null' | b: 'B'
picocli.CommandLine$ParameterException: Unexpected a or b!!!
	at net.caimanio.shell.commands.examples.Reuse$ReuseCmd.call(Reuse.java:51)
	at net.caimanio.shell.commands.examples.Reuse$ReuseCmd.call(Reuse.java:24)

I would expect that option b is null again when calling without options.

This test has different result when:
* Using defaultValue = "_NULL" ---> it works for me.
* Removing the parameter
* Adding a configuration file with defaults (but for the first time call -without options- the defaults set in the file seems to be ignored)

I didn't test more, but I think this has to be checked. Here is an example to reproduce (in JLINE!):

```
import java.util.List;
import java.util.concurrent.Callable;
import picocli.CommandLine.ArgGroup;
import picocli.CommandLine.Command;
import picocli.CommandLine.Model.CommandSpec;
import picocli.CommandLine.Option;
import picocli.CommandLine.ParameterException;
import picocli.CommandLine.Parameters;
import picocli.CommandLine.Spec;

// Class with two static classes: command and completer.
public class Reuse {

  @Command(name = "reuse", hidden = true)
  public static class ReuseCmd implements Callable {

    @Spec
    protected CommandSpec cmdSpec;

    @ArgGroup(exclusive = true)
    ExclusiveArgs exclusiveArgs;

    static class ExclusiveArgs {
      @Option(names = "-a") //, defaultValue = "_NULL_")
      static String a;

      @Option(names = "-b") //, defaultValue = "_NULL_")
      static String b;
    }

    @Parameters
    String param;

    public Integer call() {
      List<String> orgArgs = cmdSpec.commandLine().getParseResult().originalArgs();
      System.out.printf("Executing line: '%s'\n", String.join(" ", orgArgs));
      System.out.printf("a: '%s' | b: '%s'\n", ExclusiveArgs.a, ExclusiveArgs.b);

      if (orgArgs.size() == 2) {
        if (ExclusiveArgs.a != null || ExclusiveArgs.b != null) {
          throw new ParameterException(cmdSpec.commandLine(), "Unexpected a or b!!!");
        }
      }
      return 0;
    }
  }
}
```

**Comment by remkop** [comment]
Hi @eqvis, thank you for raising this!

Small detail first: I would recommend using the constant [`picocli.CommandLine.Option#NULL_VALUE`](https://picocli.info/apidocs-all/info.picocli/picocli/CommandLine.Option.html#NULL_VALUE) instead of the literal `_NULL_`.

About defaulting in argument groups: 
I agree this is a tricky area, and has come up in previous issues (https://github.com/remkop/picocli/issues/1536, https://github.com/remkop/picocli/issues/1409, maybe more).

The user manual now has a section dedicated to [Default Values in Argument Groups](https://picocli.info/#_default_values_in_argument_groups), with subsections relating to showing default values in the usage help and assigning default values to the actual fields. Hopefully that section is useful.

Looking at your example, I see that the fields for options `-a` and `-b` are declared as `static`.
This means that the same field is shared for all instances of the enclosing `ExclusiveArgs` class.
I don't think that this is your intention. I would recommend making the `String a;` and `String b;` fields non-static.

Secondly, you can try giving the `@ArgGroup`-annotated field an initial value:

```java
    // instantiating the group in the declaration:
    // all options in this group now also have their initial (default) value
    @ArgGroup(exclusive = true)
    ExclusiveArgs exclusiveArgs = new ExclusiveArgs();
```

The combination of those two changes should change the behaviour you are seeing (where the group members retained the value of the previous call when the group is not specified on the command line).

**Comment by eqvis** [comment]
Hello @remkop,

thank you for your fast answer!

I'm very sorry to tell you, that removing static from all the objects and classes and adding the initialization does not make a difference! I removed the static keyword and moved the class "ExclusiveArgs" to an other file, the command has no more static either nor have the Strings.

Only adding `defaultValue = Option.NULL_VALUE` does the thing!

Question aside: I thought this was needed by picocli, but I understood, that the cause of all my static was that I had innerclasses that I used in the main()-function. Refactoring has helped here. But is it correct, that picocli hold one instance of each command and reuses this object each call (in jline)? In this case, having static commands or not makes no difference. 
I'm thinking about a "reset-notification" when the user presses ENTER... Well perhaps moving to graalvm and removing jline would be the solution for a lot of problems here :)

 Results:
   - Non-static classes + Non-static variables + No default + No defaultValue => ERROR
   - Non-static classes + Non-static variables + default=null + No defaultValue => ERROR
   - Non-static classes + Non-static variables + No default + defaultValue="_NULL_" => OK!

Well I guess picocli has no way to know what to set as default if there is no defaultValue specified, because the java-default-value only works while initialization, right? But perhaps picocli could use this value (perhaps null) as defaultValue if there was nothing specified otherwise the behaviour is as we saw it in my tests.

Best regards,
Mayra
